lightstep/sdk/metric/internal/viewstate/collectors.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/data"
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/number"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+// compiledSyncBase is any synchronous instrument view.
+type compiledSyncBase[N number.Any, Storage any, Methods aggregator.Methods[N, Storage]] struct {
+	instrumentBase[N, Storage, Methods]
+}
+
+// NewAccumulator returns a Accumulator for a synchronous instrument view.
+func (csv *compiledSyncBase[N, Storage, Methods]) NewAccumulator(kvs attribute.Set) Accumulator {
+	sc := &syncAccumulator[N, Storage, Methods]{}
+	csv.initStorage(&sc.current)
+	csv.initStorage(&sc.snapshot)
+
+	sc.findStorage = csv.storageFinder(kvs)
+
+	return sc
+}
+
+// compiledSyncBase is any asynchronous instrument view.
+type compiledAsyncBase[N number.Any, Storage any, Methods aggregator.Methods[N, Storage]] struct {
+	instrumentBase[N, Storage, Methods]
+}
+
+// NewAccumulator returns a Accumulator for an asynchronous instrument view.
+func (cav *compiledAsyncBase[N, Storage, Methods]) NewAccumulator(kvs attribute.Set) Accumulator {
+	ac := &asyncAccumulator[N, Storage, Methods]{}
+
+	cav.initStorage(&ac.snapshot)
+	ac.current = 0
+	ac.findStorage = cav.storageFinder(kvs)
+
+	return ac
+}
+
+// statefulSyncInstrument is a synchronous instrument that maintains cumulative state.
+type statefulSyncInstrument[N number.Any, Storage any, Methods aggregator.Methods[N, Storage]] struct {
+	compiledSyncBase[N, Storage, Methods]
+}
+
+// Collect for synchronous cumulative temporality.
+func (p *statefulSyncInstrument[N, Storage, Methods]) Collect(seq data.Sequence, output *[]data.Instrument) {
+	var methods Methods
+
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	ioutput := p.appendInstrument(output)
+
+	for set, storage := range p.data {
+		p.appendPoint(ioutput, set, methods.ToAggregation(storage), aggregation.CumulativeTemporality, seq.Start, seq.Now)
+	}
+}
+
+// statelessSyncInstrument is a synchronous instrument that maintains no state.
+type statelessSyncInstrument[N number.Any, Storage any, Methods aggregator.Methods[N, Storage]] struct {
+	compiledSyncBase[N, Storage, Methods]
+}
+
+// Collect for synchronous delta temporality.
+func (p *statelessSyncInstrument[N, Storage, Methods]) Collect(seq data.Sequence, output *[]data.Instrument) {
+	var methods Methods
+
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	ioutput := p.appendInstrument(output)
+
+	for set, storage := range p.data {
+		if !methods.HasChange(storage) {
+			delete(p.data, set)
+			continue
+		}
+
+		// Possibly re-use the underlying storage.  For
+		// synchronous instruments, where accumulation happens
+		// between collection events (e.g., due to other
+		// readers collecting), we must reset the storage now
+		// or completely clear the map.
+		point, exists := p.appendOrReusePoint(ioutput)
+		if exists == nil {
+			exists = p.newStorage()
+		} else {
+			methods.Reset(exists)
+		}
+		methods.Merge(exists, storage)
```

Comment:
Yes, this is a good point. I would say a histogram `Swap()` method might be best: a call to `methods.Swap(exists, storage)` would just swap the state fields, since `exists` is already reset.

Reply:
I added a comment explaining this; I'd be glad to do this later.